### PR TITLE
[expo-updates] Add timeout info to reloadAsync() API doc

### PR DESCRIPTION
**Pull request by douglowder**
# Why

In some difficult to reproduce production cases, and in the Detox environment, `reloadAsync()` can fail if called directly. The issue seems to be that there is some running JS code that needs to complete in order for the reload to complete gracefully.

# How

~Add an optional `timeout` parameter to `Updates.reloadAsync()` API. If the timeout is greater than 0, the call to `reloadAsync()` will be wrapped in a setTimeout with that value. If no timeout is passed in, or the timeout is 0, the method will be called directly (existing behavior).~

Add documentation on the issue to the `reloadAsync()` API doc. 

# Test Plan

E2E tests have been slightly changed to note the issue. These should pass.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/src/Updates.ts [hunk 1]
```diff
@@ -142,7 +146,7 @@ const manualUpdatesInstructions = isUsingExpoDevelopmentClient
  * `ReactApplication` you want to reload, or call `UpdatesController.setReactNativeHost` with the
  * proper instance of `ReactNativeHost`.
  */
-export async function reloadAsync(): Promise<void> {
+export async function reloadAsync(timeout?: number): Promise<void> {
```
Comment (wschurman) [inline packages/expo-updates/src/Updates.ts:149]:
```suggestion
export async function reloadAsync(delay?: number): Promise<void> {
```

**Review by wschurman — COMMENTED** [review]
> This is useful for cases where it is desired to allow some long-running code to complete before reloading the app.
 
Curious how this works. Does this only work if the long-running code takes shorter than the timeout? Or does it work because it no longer awaits? Or does the await fail since it is running in the first JS instance?

At first glance, I think we need to understand the problem a bit more deeply before we add a new parameter to the API.

**Comment by douglowder** [comment]
> Curious how this works. Does this only work if the long-running code takes shorter than the timeout? Or does it work because it no longer awaits? Or does the await fail since it is running in the first JS instance?

@wschurman in local Detox testing, it appeared that the timeout had to be at least 500 ms -- a timeout of 1 or 10 ms was not sufficient. I will do some further testing and debugging to see if I can get any more information.

File: packages/expo-updates/src/Updates.ts [hunk 2]
```diff
@@ -133,6 +133,10 @@ const manualUpdatesInstructions = isUsingExpoDevelopmentClient
  * This method cannot be used in development mode, and the returned promise will be rejected if you
  * try to do so.
  *
+ * @param timeout If this is passed in, the reload instruction will be wrapped in a setTimeout() with a delay of
+ * (timeout) milliseconds. This is useful for cases where it is desired to allow some long-running code
```
Comment (douglowder) [inline packages/expo-updates/src/Updates.ts:137]:
```suggestion
 * @param delay If this is passed in, the reload instruction will be wrapped in a setTimeout() with a delay of
 * (delay) milliseconds. This is useful for cases where it is desired to allow some long-running code
```

File: packages/expo-updates/src/Updates.ts [hunk 3]
```diff
@@ -130,6 +130,10 @@ const manualUpdatesInstructions = isUsingExpoDevelopmentClient
  * executed after the `Updates.reloadAsync` method call resolves, since that depends on the OS and
  * the state of the native module and main threads.
  *
+ * In some production use cases, and in the Detox environment, reloadAsync() can
+ * fail or hang if called directly. For these use cases, this API can be called within a setTimeout(),
+ * to allow other JS and native code to complete before the app reloads.
```
Comment (ide) [inline packages/expo-updates/src/Updates.ts:135]: This guidance is too ambiguous and makes people feel they can't rely on the API. At minimum, we should describe this as a known issue we're working to resolve.

If a setTimeout call fixes this, that likely means there's something that Updates.reloadAsync() depends on that we're not waiting for. I think we need to figure out what that something is, and change the internal implementation of Updates.reloadAsync() to conceptually look like:
```
await preconditionWeDontYetKnow;
...current logic to reload
```

**Review by ide — CHANGES_REQUESTED** [review]

**Pull request closed without merging**